pulser/simulation/simulation.py

```diff
@@ -205,21 +211,62 @@ def _build_basis_and_op_matrices(self) -> None:
-        if global_op:
-            return sum(self._build_operator(op_id, q_id)
-                       for q_id in self._qdict)
-        if len(set(qubit_ids)) < len(qubit_ids):
-            raise ValueError("Duplicate atom ids in argument list.")
-        # List of identity operators, except for op_id where requested:
-        op_list = [self.op_matrix[op_id]
-                   if j in map(self._qid_index.get, qubit_ids)
-                   else self.op_matrix['I'] for j in range(self._size)]
+    def build_operator(self, operations: Union[list, tuple]) -> qutip.Qobj:
+        """Creates an operator with non trivial actions on some qubits.
+
+        Takes as argument a list of tuples [(operator_1, qubits_1),
+        (operator_2, qubits_2)...]. Returns the operator given by the tensor
+        product of {operator_i applied on qubits_i} and Id on the rest.
+        (operator, 'global') returns the sum for all $j$ of operator
+        applied at qubit $j$ and identity elsewhere.
+
+        Example for 4 qubits: [(Z, [1, 2]), (Y, [3])] returns ZZYI
+        and [(X, 'global')] returns XIII + IXII + IIXI + IIIX
+
+        Args:
+            operations (list): List of tuples (operator, qubits)
+                operator can be a qutip.Quobj or a string key for
+                self.op_matrix qubits is the list on which operator
+                will be applied. The qubits can be passed as their
+                index or their label in the register.
+
+        Returns:
+            qutip.Qobj: the final operator.
+        """
+        op_list = [self.op_matrix['I'] for j in range(self._size)]
+
+        if not isinstance(operations, list):
+            operations = [operations]
+
+        for operator, qubits in operations:
+            if qubits == 'global':
+                return sum(self.build_operator([(operator, [q_id])])
+                           for q_id in self._qdict)
+            else:
+                if len(set(qubits)) < len(qubits):
+                    raise ValueError("Duplicate atom ids in argument list.")
+                if isinstance(operator, str):
+                    if operator not in self.op_matrix.keys():
+                        raise ValueError(
+                            f"{operator} is not a valid operator")
+                    operator = self.op_matrix[operator]
+                for qubit in qubits:
+                    if isinstance(qubit, int):
+                        if qubit >= self._size:
+                            raise ValueError(
+                                "The index value is greater than the "
+                                "size of the system")
+                        k = qubit
+                    else:
+                        if qubit not in self._qid_index.keys():
+                            raise ValueError(
+                                f"{qubit} is not a valid qubit")
+                        k = self._qid_index[qubit]
```

Comment:
Maybe I'm missing something here, but can't you simplify this by doing:
```suggestion
                    try:
                        k = self._qid_index[qubit]
                    except KeyError:
                        raise ValueError(f"{qubit} is not a valid qubit")
```

Reply:
Not exactly because qubit can either be an index, or a key for the dictionnary. We can however put your suggestion in the `else`

Reply:
The thing is that it can be a key for the dictionary while also being an `int`, one does not exclude the other. You can have the indices of your qubits be `0`, `1`, ...

Reply:
If you create a Register with these qubits: `{1: (0,0), 2: (1, 1)}` what you have here will actually be wrong, because `k` for `qubit = 1` is `0` and `k` for `qubit=2` is `1`. This information will still be correct in your `_qid_index` however, which in this case will be `{1: 0, 2: 1}`. You should also use the `_qid_index` to get the value of `k`.

Reply:
What I'm getting at is that it is ambiguous to accept both qubit ids and indices, because of this situation where an `int` can be a `qid`, and you have no way of knowing. I would rather avoid this ambiguity altogether by accepting only one type.

Reply:
True. I guess that would be the responsibility of the user not to mess up. What do you think @sebgrijalva ?

Reply:
I guess keeping the qubit ids is more important since this is what the user chose and can see on the register. In a complicated graph, she will have to check `register.draw()` carefully to assign local operators.

Reply:
My vote goes to having just the qubit ids, they're the only ones being used here anyway right?